### DISCUSSION TITLE
Dudas con notificaciones de pago.

### DISCUSSION THREAD

**Discussion by coriaweb** [creagia/laravel-redsys#70] [Q&A] [answered]
Hola!

Me he vuelto a poner de nuevo con este tema y estoy un poco atascado...

A ver, si realizo una solicitud sin modelo, algo como asi:

```
$redsysRequest = RequestBuilder::newRequest(new RequestParameters(
                amountInCents: $invoice->total * 100,
                order: $invoice->id,
                currency: Currency::EUR,
                transactionType: TransactionType::Autorizacion,
                productDescription: 'Pago de factura ' . $invoice->id ,
                excepSca: 'MIT',
                merchantUrl: route('redsys.notification'),
            ))->usingCard(CofType::Recurring, $redsysCard);

            // Redirigir al usuario a Redsys
            return $redsysRequest->redirect();
```
            
 en la ruta de redsys.notificacion he creado esto:
 
 ```
namespace App\Http\Controllers;

use Creagia\Redsys\Enums\CofType;
use Creagia\Redsys\Enums\PayMethod;
use Illuminate\Support\Facades\Auth;
use Illuminate\Http\Request;
use Creagia\Redsys\Facades\Redsys;
use Creagia\Redsys\RedsysResponse;
use Illuminate\Support\Facades\Log;

class RedsysController extends Controller
{

    // Esta función maneja la notificación desde Redsys
    public function notification(Request $request)
    {
    Log::info('Método notification ejecutado.');
    }
}
```

cuando con la plataforma local hago cualquier notificacion, pago realizado o cualquier error, no se notifica. ¿Que estoy haciendo mal?

**Comment by dtorras** [comment]
Hola. Cuando dices que lo usas en local, te refieres a que estás usando el ["local gateway"](https://github.com/creagia/laravel-redsys?tab=readme-ov-file#local-gateway)? Si es así, comprueba que tengas actualizado el paquete a la versión 3.x. En las versiones anteriores no funcionaba la notificación local si se usaba Laravel Sail o `artisan serve`, solo funcionaba con Laravel Valet o otros entornos con nginx o Apache. Aquí el cambio: https://github.com/creagia/laravel-redsys/pull/63

Por otra parte, te recomiendo que uses el paquete relacionado los pagos de Redsys con un modelo Eloquent. Si no lo relacionas, este paquete no te aporta mucho y tendrías más flexibilidad usando el paquete base `redsys-php` https://github.com/creagia/redsys-php

**Reply by coriaweb** [reply]
Hola!

Si, uso de momento local gateway. El problema es que me viene bien "a medias", es decir, hay cosas que estoy vinculando a un modelo y funciona bien, pero hay otras que necesito poder tramitar yo la notificación y hacer algo yo directamente.

Si al código le añado algo asi:

merchantUrl: route('mi-notificacion.redsys'),

no funciona, parece que no llega a la url indicada nada, el log: 

Log::info('Método notification ejecutado.');

no se ejecuta, si no lo añado, usa el default del paquete y si funciona, pero no se donde se tramita jejeje

Voy a ver el paquete que dices, sirve también para laravel verdad? ¿Se puede usar ambos? Este para los relacionados y el otro para los que van "por libre"? 

Saludos y muchas gracias por molestarte en responder.

**Reply by dtorras** [reply, marked as answer]
Creo que lo mejor puede ser que uses el default del paquete para gestionar las notificaciones. Puedes crear un event listener de las notificaciones para los pagos que no están relacionados con el modelo: https://github.com/creagia/laravel-redsys?tab=readme-ov-file#events

Con el [RedsysSuccessfulEvent](https://github.com/creagia/laravel-redsys?tab=readme-ov-file#redsyssuccessfulevent) y el [RedsysUnsuccessfulEvent](https://github.com/creagia/laravel-redsys?tab=readme-ov-file#redsysunsuccessfulevent), puedes gestionar los resultados.

La app de ejemplo está un poco desactualizada pero tiene una implementación de un `RedsysSuccessfulListener`: https://github.com/creagia/laravel-redsys-example/blob/main/app/Providers/EventServiceProvider.php

**Reply by coriaweb** [reply]
Me da error 404: https://github.com/creagia/laravel-redsys-example/blob/main/app/Providers/EventServiceProvider.php

Si tienes algún ejemplo de como hacerlo.. me haces un hombre jejeje

**Reply by dtorras** [reply]
Perdona, pensaba que el repositorio era público. Era un ejemplo de implementar [un listener](https://laravel.com/docs/10.x/events) básico en Laravel.

app/Providers/EventServiceProvider.php
```php
<?php

namespace App\Providers;

use App\Listeners\RedsysNotificationListener;
use App\Listeners\RedsysSuccessfulListener;
use Creagia\LaravelRedsys\Events\RedsysNotificationEvent;
use Creagia\LaravelRedsys\Events\RedsysSuccessfulEvent;
use Illuminate\Auth\Events\Registered;
use Illuminate\Auth\Listeners\SendEmailVerificationNotification;
use Illuminate\Foundation\Support\Providers\EventServiceProvider as ServiceProvider;
use Illuminate\Support\Facades\Event;

class EventServiceProvider extends ServiceProvider
{
    /**
     * The event listener mappings for the application.
     *
     * @var array<class-string, array<int, class-string>>
     */
    protected $listen = [
        Registered::class => [
            SendEmailVerificationNotification::class,
        ],
        RedsysNotificationEvent::class => [
            RedsysNotificationListener::class,
        ],
        RedsysSuccessfulEvent::class => [
            RedsysSuccessfulListener::class,
        ],
    ];

    /**
     * Register any events for your application.
     *
     * @return void
     */
    public function boot()
    {
        //
    }
}
```

app/Listeners/RedsysSuccessfulListener.php
```php
<?php

namespace App\Listeners;

use Creagia\LaravelRedsys\Events\RedsysSuccessfulEvent;
use Illuminate\Support\Facades\Log;

class RedsysSuccessfulListener
{
    /**
     * Create the event listener.
     *
     * @return void
     */
    public function __construct()
    {
        //
    }

    /**
     * Handle the event.
     *
     * @param  object  $event
     * @return void
     */
    public function handle(RedsysSuccessfulEvent $event)
    {
        Log::debug($event->notificationData);
    }
}
```

**Reply by coriaweb** [reply]
Funciona! jejeje eres un artista! :D

Solo una cosa... ¿Porque me duplica?

https://i.gyazo.com/37aff652ca5c287189d8988e5834b8bf.png

¿Qué estoy haciendo mal? solo tengo esto:

```
<?php

namespace App\Listeners;

use Creagia\LaravelRedsys\Events\RedsysSuccessfulEvent;
use Illuminate\Support\Facades\Log;

class RedsysSuccessfulListener
{
    /**
     * Create the event listener.
     *
     * @return void
     */
    public function __construct()
    {
        //
    }

    /**
     * Handle the event.
     *
     * @param  object  $event
     * @return void
     */
    public function handle(RedsysSuccessfulEvent $event)
    {
        Log::info('Método notification ejecutado.');
    }
}
```

**Reply by coriaweb** [reply]
Te confirmo que siempre se duplica, he añadido otro e igual:

https://i.gyazo.com/a5b084f8fad9abb12a7e4ab441c8fd53.png

Alguna idea de donde la puedo estar liando? jejeje

**Reply by dtorras** [reply]
Tendrías que hacer el debug en tu aplicación. Si encuentras algún problema concreto en el package, no dudes en abrir un issue.

**Reply by coriaweb** [reply]
Voy a probar y ver porque lo duplica, gracias!! A ver si doy con la tecla.

**Reply by coriaweb** [reply]
Solucionado!

Mientras me respondías ya estaba revisando antes, jajaja, te digo, he eliminado todo de app/Providers/EventServiceProvider.php, simplemente crear los listeners:

```
PS C:\xamppphp8.2\htdocs\*****> php artisan make:listener RedsysNotificationListener   

   INFO  Listener [C:\xamppphp8.2\htdocs\*****\app\Listeners\RedsysNotificationListener.php] created successfully.

PS C:\xamppphp8.2\htdocs\*****> php artisan make:listener RedsysSuccessfulListener

   INFO  Listener [C:\xamppphp8.2\htdocs\*****\app\Listeners\RedsysSuccessfulListener.php] created successfully.

PS C:\xamppphp8.2\htdocs\****> php artisan make:listener RedsysUnsuccessfulListener

   INFO  Listener [C:\xamppphp8.2\htdocs\****\app\Listeners\RedsysUnsuccessfulListener.php] created successfully.
```
   
   y dentro de cada uno, algo similar a esto:
   
```
   <?php

namespace App\Listeners;

use Creagia\LaravelRedsys\Events\RedsysUnsuccessfulEvent;
use Illuminate\Support\Facades\Log;

class RedsysUnsuccessfulListener
{
    /**
     * Create the event listener.
     */
    public function __construct()
    {
        //
    }

    /**
     * Handle the event.
     */
    public function handle(RedsysUnsuccessfulEvent $errorMessage)
    {
        Log::info('Método Error ejecutado.'.json_encode($errorMessage));
    }
}
```

cada uno con sus respectivos valores... y funciona bien!! no hay que volverlos a especificar en EventServiceProvider.php jejeje, te lo dejo todo por si te sirve para "documentarlo". No se si esta bien o mal, pero funciona 👯‍♂️ Ahora ya me encargo yo de tramitar los valores según vengan.

Saludos y gracias!!!
